qemu/tests/uefishell.py

```diff
@@ -0,0 +1,152 @@
+        """
+        params = self.params
+        ovmf_path = params["ovmf_path"]
+        uefishell_filename = "UefiShell.iso"
+        uefishell_dst_path = "images/%s" % uefishell_filename
+        params["cdroms"] = "%s %s" % (
+            params["cdroms"], "uefishell")
+        params["cdrom_uefishell"] = uefishell_dst_path
+        uefishell_src_path = utils_misc.get_path(
+            ovmf_path, uefishell_filename)
+        uefishell_dst_path = utils_misc.get_path(
+            data_dir.get_data_dir(), uefishell_dst_path)
+        if not os.path.exists(uefishell_dst_path):
+            cp_command = "cp -f %s %s" % (
+                uefishell_src_path, uefishell_dst_path)
+            process.system(cp_command)
+        params["bootindex_uefishell"] = "0"
+        params["start_vm"] = "yes"
+        params["shell_prompt"] = r"(Shell|FS\d:\\.*)>"
+        params["shell_linesep"] = r"\r\n"
+        env_process.process(self.test, params, self.env,
+                            env_process.preprocess_image,
+                            env_process.preprocess_vm)
+        # Backup the VAR file for the vm if need, since the operations
+        # inside uefishell will be recorded into VAR file and may
+        # affact other test cases.
+
+        if params.get("var_backup") == "yes":
+            self.var_backup()
+        vm = self.env.get_vm(params["main_vm"])
+        self.session = vm.wait_for_serial_login()
+        if under_fs0 == "yes":
+            self.send_command("fs0:")
+
+    def var_backup(self, restore=False):
+        """
+        Backup or restore the VAR file.
+
+        :param restore: Backup file if False, and restore it if True
+        """
+        path = storage.get_image_filename_filesytem(
+            self.params, data_dir.get_data_dir())
+        vars_path = "%s.fd" % path
+        if os.path.exists(vars_path):
+            if restore:
+                logging.debug("Restore var %s from %s.backup", vars_path, vars_path)
+                cmd = "cp -f %s.backup %s" % (vars_path, vars_path)
+            else:
+                logging.debug("Backup var %s to %s.backup", vars_path, vars_path)
+                cmd = "cp -f %s %s.backup" % (vars_path, vars_path)
+            process.system(cmd)
+
+    def send_command(self, command, check_result=False):
+        """
+        Send a command line to uefi shell, and check the output
+        if 'check_result' exists, and fail the case if the output
+        does not meet the expectation
+
+        :param command: the command string being executed
+        :param check_result: the pattern to validate output
+        """
+        logging.info("Send uefishell command: %s" % command)
+        output = self.session.cmd_output_safe(command)
+        if check_result:
+            if not re.search(check_result, output):
+                self.test.fail("The command result is: %s, which does not"
+                               " match the expectation: %s" % output, check_result)
```

Comment:
03:04:10 ERROR| Traceback (most recent call last):
03:04:10 ERROR|   File "/home/tp-qemu/qemu/tests/uefishell.py", line 152, in run
03:04:10 ERROR|     uefishell_test.send_command(command, check_result)
03:04:10 ERROR|   File "/home/tp-qemu/qemu/tests/uefishell.py", line 106, in send_command
03:04:10 ERROR|     " match the expectation: %s" % output, check_result)
03:04:10 ERROR| TypeError: not enough arguments for format string

It should be "% (output, check_result)".